modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/internal/application/MuleApplicationClassLoader.java

```diff
@@ -58,14 +60,36 @@ protected String findLibrary(String name) {
     return new String[] {toFile(classLoaderModel.getUrls()[0]).getPath()};
   }
 
+
+  /**
+   * Resolves the plugin classloader of the thread context class loader artifact and all it's ancestor
+   */
+  public static List<ClassLoader> resolveContextArtifactPluginClassLoaders() {
+    Set<ClassLoader> resolvedClassLoaders = new HashSet<>();
+    ClassLoader originalClassLoader = currentThread().getContextClassLoader();
+    ClassLoader tmpClassLoader = originalClassLoader;
+    resolvedClassLoaders.addAll(resolveContextArtifactPluginClassLoadersForCurrentClassLoader());
+    try {
+      while (tmpClassLoader.getParent() != null
+          && MuleDeployableArtifactClassLoader.class.isAssignableFrom(tmpClassLoader.getParent().getClass())) {
+        tmpClassLoader = tmpClassLoader.getParent();
+        currentThread().setContextClassLoader(tmpClassLoader);
+        resolvedClassLoaders.addAll(resolveContextArtifactPluginClassLoadersForCurrentClassLoader());
+      }
+    } finally {
```

Comment:
technically, the finally block must wrap the part when the change is done. (ie: on every change)